numba/cuda/compiler.py

```diff
@@ -589,16 +595,25 @@ def _prepare_args(self, ty, val, stream, retr, kernelargs):
         """
         Convert arguments to ctypes and append to kernelargs
         """
+
+        # map the arguments using any extension you've registered
+        ty, val = six.moves.reduce(
```

Comment:
Why use `reduce`?  All extensions will be applied from left to right.  Given the `nocopy` decorator is appending new extensions, I think extensions should be applied right-to-left.  Also, can you document what extension should do if it should fallthrough